pkg/v1/partial/index.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package partial
+
+import (
+	"bytes"
+	"fmt"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+var defaultPlatform = v1.Platform{
+	Architecture: "amd64",
+	OS:           "linux",
+}
+
+// WithRawIndexManifest defines the subset of v1.ImageIndex used by ImageByArch.
+type WithRawIndexManifest interface {
+	RawManifest() ([]byte, error)
```

Comment:
I'm actually concerned that this is too general because a `v1.Image` that also implements `Image(Hash)` will satisfy this, but will not work at all because `v1.Image`'s descriptors are under `Layers`. Should we just have `ImageByPlatform` take a `v1.ImageIndex` instead?